rosidl_parser/rosidl_parser/definition.py

```diff
@@ -107,170 +166,233 @@ def __eq__(self, other):
         super().__init__()
         self.namespaces = namespaces
         self.name = name
 
+    def namespaced_name(self) -> Tuple[str, ...]:
+        return (*self.namespaces, self.name)
+
     def __eq__(self, other):
         return super().__eq__(other) and \
             self.namespaces == other.namespaces and self.name == other.name
 
 
-class BaseString(BaseType):
-    """The base class of string types."""
+class AbstractGenericString(AbstractNestableType):
+    """The abstract base class of all string types."""
+
+    __slots__ = ()
+
+    def has_maximum_size(self):
+        raise NotImplementedError('Only implemented in subclasses')
+
+
+class AbstractString(AbstractGenericString):
+    """The abstract base class of 8-bit string types."""
+
+    __slots__ = ()
+
+
+class BoundedString(AbstractString):
+    """A 8-bit string type with a limited number of characters."""
 
     __slots__ = ('maximum_size', )
 
-    def __init__(self, maximum_size=None):
+    def __init__(self, maximum_size: int):
         """
         Constructor.
 
-        :param int maximum_size: the maximum length of the string in
-          characters, or None if unlimited
+        :param maximum_size: the maximum length of the string in characters
+          (must be greater than zero)
         """
+        super().__init__()
+        assert maximum_size > 0
         self.maximum_size = maximum_size
 
+    def has_maximum_size(self):
+        return True
+
     def __eq__(self, other):
         return super().__eq__(other) and \
             self.maximum_size == other.maximum_size
 
 
-class String(BaseString):
-    """A 8-bit string type."""
+class UnboundedString(AbstractString):
+    """A 8-bit string type with an unlimited number of characters."""
 
     __slots__ = ()
 
-    def __init__(self, maximum_size=None):
-        """
-        Constructor.
+    def has_maximum_size(self):
+        return False
 
-        :param int maximum_size: the maximum length of the string in
-          characters, or None if unlimited
-        """
-        super().__init__(maximum_size=maximum_size)
+
+class AbstractWString(AbstractGenericString):
+    """The abstract base class of 16-bit string types."""
+
+    __slots__ = ()
 
 
-class WString(BaseString):
+class BoundedWString(AbstractWString):
     """A 16-bit string type."""
 
-    __slots__ = ()
+    __slots__ = ('maximum_size', )
 
-    def __init__(self, maximum_size=None):
+    def __init__(self, maximum_size: int):
         """
         Constructor.
 
-        :param int maximum_size: the maximum length of the string in
-          characters, or None if unlimited
+        :param maximum_size: the maximum length of the string in characters
+          (must be greater than zero)
         """
-        super().__init__(maximum_size=maximum_size)
+        super().__init__()
+        # TODO(dirk-thomas) can't be enforced yet since the parser might pass a
+        # constant name
+        # assert maximum_size > 0
```

Comment:
While I wanted to add this `assert` it turns out the value is currently also being used to temporary store a constant name during the parsing which gets replaced with the constant value shortly after.

I won't change this during this renaming but added a note to be aware of it.